File: react_vite/src/core/authSession.ts
/**
 * authSession.ts
 * Holds the tenant-scoped JWT minted by the `auth` Edge Function and drives the
 * login / register / refresh calls.
 *
 * This replaces the old scheme where the client queried admin_users directly
 * and compared the password to password_hash. Nothing here ever sees a stored
 * password, and every Supabase request now travels as an `authenticated` role
 * carrying a tenant_id claim, which is what the RLS policies filter on.
 */
import { FUNCTIONS_URL, SUPABASE_ANON_KEY } from './supabaseConfig';

const K_TOKEN = 'auth_token';
const K_REFRESH = 'auth_refresh_token';
const K_EXPIRES = 'auth_token_expires_at';
const K_USER_ID = 'auth_user_id';

/**
 * Supabase access tokens last about an hour, so this is a small margin before
 * expiry rather than the long window a self-minted token allowed. The refresh
 * token is what actually persists the session across months of uptime.
 */
const REFRESH_WINDOW_MS = 5 * 60 * 1000; // 5 minutes

export interface AuthSession {
  token: string;
  refreshToken: string;
  expiresAt: number;
  userId: string;
  tenantId: string;
  username: string;
  mobile: string;
  email: string;
  mosqueName: string;
}

let cachedToken: string | null = null;

export const AuthSession = {
  getToken(): string | null {
    if (cachedToken) return cachedToken;
    cachedToken = localStorage.getItem(K_TOKEN);
    return cachedToken;
  },

  getExpiresAt(): number {
    return Number(localStorage.getItem(K_EXPIRES) ?? 0);
  },

  isExpired(): boolean {
    const exp = AuthSession.getExpiresAt();
    return exp > 0 && Date.now() >= exp;
  },

  getRefreshToken(): string | null {
    return localStorage.getItem(K_REFRESH);
  },

  save(session: AuthSession): void {
    cachedToken = session.token;
    localStorage.setItem(K_TOKEN, session.token);
    if (session.refreshToken) localStorage.setItem(K_REFRESH, session.refreshToken);
    localStorage.setItem(K_EXPIRES, String(session.expiresAt));
    localStorage.setItem(K_USER_ID, session.userId);
  },

  clear(): void {
    cachedToken = null;
    localStorage.removeItem(K_TOKEN);
    localStorage.removeItem(K_REFRESH);
    localStorage.removeItem(K_EXPIRES);
    localStorage.removeItem(K_USER_ID);
  },

  async login(identifier: string, password: string): Promise<AuthSession> {
    const session = await callAuth({ action: 'login', identifier, password });
    AuthSession.save(session);
    return session;
  },

  async register(params: {
    mosqueName: string;
    username: string;
    password: string;
    mobile?: string;
    email?: string;
  }): Promise<AuthSession> {
    const session = await callAuth({ action: 'register', ...params });
    AuthSession.save(session);
    return session;
  },

  /**
   * Renews the access token when it is close to expiring. Returns false only
   * when the session is genuinely gone and the user has to sign in again; a
   * network failure leaves the existing token in place so a display that is
   * merely offline does not log itself out.
   */
  async refreshIfNeeded(): Promise<boolean> {
    const token = AuthSession.getToken();
    const refreshToken = AuthSession.getRefreshToken();
    if (!token && !refreshToken) return false;

    const expiresAt = AuthSession.getExpiresAt();
    if (token && expiresAt > 0 && expiresAt - Date.now() > REFRESH_WINDOW_MS) return true;

    // Nothing to renew with: the token is stale and cannot be replaced.
    if (!refreshToken) return false;

    try {
      const session = await callAuth({ action: 'refresh', refreshToken });
      AuthSession.save(session);
      return true;
    } catch (e) {
      console.warn('[Auth] refresh failed', e);
      return !AuthSession.isExpired();
    }
  },
};

async function callAuth(body: Record<string, unknown>): Promise<AuthSession> {
  const res = await fetch(`${FUNCTIONS_URL}/auth`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      apikey: SUPABASE_ANON_KEY,
      Authorization: `Bearer ${SUPABASE_ANON_KEY}`,
    },
    body: JSON.stringify(body),
  });

  const data = await res.json().catch(() => ({}));
  if (!res.ok) throw new Error((data as { error?: string }).error ?? `Auth failed (${res.status})`);
  return data as AuthSession;
}

File: flutter_app/lib/core/auth_session.dart
/// auth_session.dart
/// Holds the tenant-scoped JWT minted by the `auth` Edge Function and drives
/// the login / register / refresh calls.
///
/// This replaces the old scheme where the client queried admin_users directly
/// and compared the typed password to password_hash — which required the anon
/// key baked into the app to have read access to the credential table. Nothing
/// here ever sees a stored password, and every Supabase request now travels as
/// an `authenticated` role carrying a tenant_id claim, which is what the RLS
/// policies in supabase/02_security_hardening.sql filter on.
library;

import 'dart:convert';
import 'dart:developer' as dev;
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class AuthSessionData {
  final String token;
  final String refreshToken;
  final int expiresAt; // epoch ms
  final String userId;
  final String tenantId;
  final String username;
  final String mobile;
  final String email;
  final String mosqueName;

  const AuthSessionData({
    required this.token,
    required this.refreshToken,
    required this.expiresAt,
    required this.userId,
    required this.tenantId,
    required this.username,
    required this.mobile,
    required this.email,
    required this.mosqueName,
  });

  factory AuthSessionData.fromJson(Map<String, dynamic> j) => AuthSessionData(
        token: j['token'] as String,
        refreshToken: j['refreshToken'] as String? ?? '',
        expiresAt: (j['expiresAt'] as num?)?.toInt() ?? 0,
        userId: j['userId'] as String? ?? '',
        tenantId: j['tenantId'] as String,
        username: j['username'] as String? ?? '',
        mobile: j['mobile'] as String? ?? '',
        email: j['email'] as String? ?? '',
        mosqueName: j['mosqueName'] as String? ?? 'Linked Mosque',
      );
}

class AuthSession {
  static const _kToken = 'auth_token';
  static const _kRefreshToken = 'auth_refresh_token';
  static const _kExpiresAt = 'auth_token_expires_at';
  static const _kUserId = 'auth_user_id';

  /// Supabase access tokens last about an hour, so this is a margin before
  /// expiry rather than a renewal interval. The refresh token is what keeps a
  /// display signed in across months of uptime.
  static const _refreshWindow = Duration(minutes: 5);

  static late SharedPreferences _prefs;
  static late String _functionsUrl;
  static late String _anonKey;
  static String? _cachedToken;

  static Future<void> init({
    required String supabaseUrl,
    required String anonKey,
  }) async {
    _prefs = await SharedPreferences.getInstance();
    _functionsUrl = '$supabaseUrl/functions/v1';
    _anonKey = anonKey;
    _cachedToken = _prefs.getString(_kToken);
  }

  /// Synchronous so it can back the client's accessToken callback cheaply.
  static String? get token => _cachedToken;

  static int get expiresAt => _prefs.getInt(_kExpiresAt) ?? 0;

  static bool get isExpired {
    final exp = expiresAt;
    return exp > 0 && DateTime.now().millisecondsSinceEpoch >= exp;
  }

  static String? get refreshToken => _prefs.getString(_kRefreshToken);

  static Future<void> _save(AuthSessionData s) async {
    _cachedToken = s.token;
    await _prefs.setString(_kToken, s.token);
    if (s.refreshToken.isNotEmpty) {
      await _prefs.setString(_kRefreshToken, s.refreshToken);
    }
    await _prefs.setInt(_kExpiresAt, s.expiresAt);
    await _prefs.setString(_kUserId, s.userId);
  }

  static Future<void> clear() async {
    _cachedToken = null;
    await _prefs.remove(_kToken);
    await _prefs.remove(_kRefreshToken);
    await _prefs.remove(_kExpiresAt);
    await _prefs.remove(_kUserId);
  }

  static Future<AuthSessionData> login(String identifier, String password) async {
    final s = await _call({
      'action': 'login',
      'identifier': identifier,
      'password': password,
    });
    await _save(s);
    return s;
  }

  static Future<AuthSessionData> register({
    required String mosqueName,
    required String username,
    required String password,
    String? mobile,
    String? email,
  }) async {
    final s = await _call({
      'action': 'register',
      'mosqueName': mosqueName,
      'username': username,
      'password': password,
      'mobile': mobile,
      'email': email,
    });
    await _save(s);
    return s;
  }

  /// Renews the access token when it is close to expiring. Returns false only
  /// when the session is genuinely gone and the user has to link again; a
  /// network failure leaves the existing token in place so a display that is
  /// merely offline does not sign itself out.
  static Future<bool> refreshIfNeeded() async {
    final current = _cachedToken;
    final refresh = refreshToken;
    if (current == null && refresh == null) return false;

    final exp = expiresAt;
    if (current != null &&
        exp > 0 &&
        exp - DateTime.now().millisecondsSinceEpoch > _refreshWindow.inMilliseconds) {
      return true;
    }

    if (refresh == null) return false;

    try {
      await _save(await _call({'action': 'refresh', 'refreshToken': refresh}));
      return true;
    } catch (e) {
      dev.log('[Auth] refresh failed: $e');
      return !isExpired;
    }
  }

  static Future<AuthSessionData> _call(Map<String, dynamic> body) async {
    final res = await http.post(
      Uri.parse('$_functionsUrl/auth'),
      headers: {
        'Content-Type': 'application/json',
        'apikey': _anonKey,
        'Authorization': 'Bearer $_anonKey',
      },
      body: jsonEncode(body),
    );

    Map<String, dynamic> data;
    try {
      data = jsonDecode(res.body) as Map<String, dynamic>;
    } catch (_) {
      data = {};
    }

    if (res.statusCode < 200 || res.statusCode >= 300) {
      throw Exception(data['error'] as String? ?? 'Auth failed (${res.statusCode})');
    }
    return AuthSessionData.fromJson(data);
  }
}
